tfx/experimental/distributed_inference/graphdef_experiments/subgraph_partitioning/graph_partition.py

```diff
@@ -0,0 +1,471 @@
+
+
+def _get_graph_def(filepath):
+  graph_def = graph_pb2.GraphDef()
+  with tf.compat.v1.gfile.FastGFile(filepath, 'rb') as f:
+    graph_def.ParseFromString(f.read())
+  return graph_def
+
+
+def partition_all_graphs(op_to_graph_def, op_to_outputs):
+  """Partition all the graph_defs.
+
+  In our example, we have three graphs: 'main', 'remote_op_a', 'remote_op_b'.
+  Here, we partition each graph given its graph_def and output names.'
+
+  Args:
+    op_to_graph_def: {a graph name: a GraphDef proto}
+    op_to_outputs: {a graph name: a list of output node names}
+
+  Returns:
+    op_to_execution_specs: {a graph name: a list of execution specs}
+      each execution spec is a dict which comprises:
+        {'subgraph': a GraphDef,
+         'inputs': a set of input node names,
+         'outputs': a set of output node names,
+         'is_remote_op': a Boolean}
+  """
+  op_to_execution_specs = {}
+  for op in op_to_graph_def:
+    execution_specs = _partition_one_graph(op_to_graph_def[op],
+                                           op_to_outputs[op])
+    op_to_execution_specs[op] = execution_specs
+  return op_to_execution_specs
+
+
+def _partition_one_graph(graph_def, outputs):
+  """Partition a graph_def.
+
+  Args:
+    graph_def: a GraphDef proto
+    outputs: a set of output node names
+
+  Returns:
+    execution_specs: a list of specs, each spec is a dict that contains:
+      {'subgraph': a GraphDef,
+       'inputs': a set of input node names,
+       'outputs': a set of output node names,
+       'is_remote_op': a Boolean}
+  """
+  graph = _get_graph(graph_def)
+  node_name_to_node_def = _get_node_name_to_node_def(graph_def)
+  node_name_to_input_names = _get_node_name_to_input_names(graph_def)
+
+  remote_op_relations = _get_remote_op_relations(graph_def,
+                                                 node_name_to_node_def,
+                                                 node_name_to_input_names)
+
+  execution_specs = _get_execution_specs(graph_def,
+                                         graph,
+                                         node_name_to_node_def,
+                                         node_name_to_input_names,
+                                         remote_op_relations,
+                                         outputs)
+  return execution_specs
+
+
+def _get_graph(graph_def):
+  temp = tf.Graph()
+  with temp.as_default():
+    tf.import_graph_def(graph_def)
+    return tf.compat.v1.get_default_graph()
+
+
+def _get_node_name_to_node_def(graph_def):
+  return {node.name: node for node in graph_def.node}
+
+
+def _get_node_name_to_input_names(graph_def):
+  return {node.name: list(node.input) for node in graph_def.node}
+
+
+def _get_remote_op_relations(graph_def,
+                             node_name_to_node_def,
+                             node_name_to_input_names):
+  """Get the execution dependencies between remote ops.
+
+  The remote op children must be executed before executing a remote op.
```

Comment:
children -> inputs / parent? This seems like the opposite of the usual meaning of "children".

Same everywhere else.

Reply:
Changed.